File: 16. 3Sum Closest/Fixed/3sum_closest.cs
using System;

namespace sum_of_three_closest
{
	class sum_of_three_closest
	{
		static int result_number = 0;
		static int number_of_elements = 3;
		
		public class Result{
			int [] values = new int [number_of_elements];
			int number_of_values;
			bool not_null;
			
			public void set_values (int [] list){
				number_of_values = list.Length;
				not_null = true;
				
				for (int i = 0; i < list.Length; i++){
					values[i] = list[i];
				}
			}
			
			public void empty_values (){
				not_null = false;
				int len_list = number_of_elements;
				
				for (int i = 0; i < len_list; i++){
					values[i] = 0;
				}
				
				number_of_values = 0;
			}
			
			public Result (int [] list){
				set_values (list);
			}
			
			
			public Result (){
			}
			
			
			public void print_values(){
				Console.WriteLine("This is the handy work of C#");
				int nje = 0;
				
				for (int i = 0; i < number_of_values; i++){
					Console.Write(values[i]);
					nje += values[i];
					
					if (i < (number_of_values - 1)){
						Console.Write("\t");
					}
				}
				
				Console.Write("\nThe sum of the values is:\t");
				Console.WriteLine(nje);
			}
		}
		
		static int factorial (int number){
			int nje = 1;
			
			for (int i = number; i > 0; i--){
				nje *= i;
			}
			
			return (nje);
		}
		
		static int absolute (int number){
			return (number < 0) ? (number * -1) : (number * 1);
		}
		
		static Result [] fun_sum_of_three_closest (int [] list, int target, int sample_size){
			int result_cursor = 0;
			int closest_difference = 0;
			bool first_execution = true;
			int len_list = list.Length;
			Result [] nje = new Result[sample_size];
			
			for (int a = 0; a < len_list; a++){
				for (int b = 0; b < len_list; b++){
					if (b > a){
						for (int c = 0; c < len_list; c++){
							if (c > b){
								if ((a != b) && (a != c) && (b != c)){
									int aa = list[a];
									int bb = list[b];
									int cc = list[c];
									
									//if ((aa != bb) && (aa != cc) && (bb != cc)){
									if (true){
										int difference = absolute((aa + bb + cc) - target);
										bool update_array = false;
										
										if (first_execution){
											closest_difference = difference;
											first_execution = false;
											
											update_array = true;
										} else {
											if (difference < closest_difference){
												closest_difference = difference;
												
												if (result_cursor > 0){
													for (int i = 0; i < result_cursor; i++){
														nje[result_cursor].empty_values();
													}
													result_cursor = 0;
												}
												
												update_array = true;
											} else if (difference == closest_difference){
												update_array = true;
											}
										}
										
										if (update_array){
											int [] y = {aa, bb, cc};
											nje[result_cursor] = new Result(y);
											result_cursor++;
											result_number = result_cursor;
										}
									}
								}
							}
						}
					}
				}
			}
			
			return (nje);
		}
		
	
		static void Main (string [] args){
			int [] nums = {-1,2,1,-4};
			int len_nums = nums.Length;
			int target = 1;
			
			int sample_space = (factorial(len_nums) / (factorial(len_nums - number_of_elements) * factorial(number_of_elements)));
			
			Result [] h = fun_sum_of_three_closest(nums, target, sample_space);
			
			for (int i = 0; i < result_number; i++){
				h[i].print_values();
			}
		}
	
	}
		
}

File: 16. 3Sum Closest/Fixed/3sum_closest.rs
static mut result_number: isize = 0;
const number_of_elements: usize = 3;
const sample_size: isize = 8;


#[derive(Copy, Clone)]
struct Result{
	values: [isize; number_of_elements],
	number_of_values: isize,
	not_null: bool
}

impl Result{
	fn new(list: &[isize]) -> Self{
		let len_list = list.len() as isize;
		let mut arr: [isize; number_of_elements] = [0; number_of_elements];
		
		for i in 0..len_list{
			arr[i as usize] = list[i as usize];
		}
		
		return Result{values: arr, number_of_values: len_list, not_null: true};
	}
	
	fn empty() -> Self{
		return Result{values: [0; number_of_elements], number_of_values: number_of_elements as isize, not_null: false};
	}
	
	fn set_values (&mut self, list: &[isize]){
		self.not_null = true;
		self.number_of_values = list.len() as isize;
		
		for i in 0..self.number_of_values{
			self.values[i as usize] = list[i as usize];
		}
	}
	
	fn print_values (self) -> (){
		println!("This is the handy work of Rust");
		
		let mut out: isize = 0;
		
		for i in 0..self.number_of_values{
			print!("{}", self.values[i as usize]);
			out += self.values[i as usize];
			
			if i < (self.number_of_values - 1){
				print!("\t");
			}
		}
		println!("\nThe sum of the values is {}\n\n", out);
	}
	
	fn empty_result (&mut self) -> (){
		self.not_null = false;
		
		for i in 0..self.number_of_values{
			self.values[i as usize] = 0;
		}
		
		self.number_of_values = 0;
	}
}

fn factorial (number: isize) -> isize{
	let mut out = 1;
	
	for i in (1..=number).rev(){
		out *= i;
	}
	
	return out;
}

fn absolute (number: isize) -> isize{
	return if (number < 0) {(number * -1)} else {(number * 1)};
}

fn sum_of_three_closest (list: &[isize], target: isize, out: &mut[Result]) -> (){
//fn sum_of_three_closest (list: &[isize], target: isize) -> Result{
	let mut result_cursor: usize = 0;
	let mut closest_difference: isize = 0;
	let mut first_execution: bool = true;
	let len_list: usize = list.len();
	//let mut out: Result;
	
	for a in 0..len_list{
		for b in 0..len_list{
			if b > a{
				for c in 0..len_list{
					if c > b{
						if (a != b) && (a != c) && (b != c){
							let aa = list[a];
							let bb = list[b];
							let cc = list[c];
							
							//if (aa != bb) && (aa != cc) && (bb != cc){
							
							if true{	
								let difference: isize = absolute((aa + bb + cc) - target);
								let mut update_array: bool = false;
								
								if first_execution{
									closest_difference = difference;
									first_execution = false;
									update_array = true;
								} else {
									if (difference < closest_difference){
										closest_difference = difference;
										/*
										//This is the one that returns the first combination
										out = Result::empty();	
										*/
									
									//BEGIN if commenting out for the first occurrence outcome
										if (result_cursor > 0){
											for i in 0..result_cursor{
												out[result_cursor].empty_result();
											}
											
											result_cursor = 0;
										}
										
										update_array = true;
									} else if (difference == closest_difference){
										update_array = true;
									//END if commenting out for the first occurrence outcome
									}
								}
								
								if update_array{
									let y: [isize; number_of_elements] = [aa, bb, cc];
									
									out[result_cursor] = Result::new(&y);
									
									result_cursor += 1;
									unsafe{
										result_number = result_cursor as isize;
									}
								}
							}
						}
					}
				}
			}
		}
	}
	
	//return out; //This is for the first occurrence
}

fn main (){
	let nums: &[isize] = &[-1,2,1,-4];
	let len_nums: usize = nums.len() as usize;
	
	let sample_space = (factorial(len_nums as isize) / (factorial((len_nums - number_of_elements) as isize) * factorial (number_of_elements as isize)));
	
	if (sample_size < sample_space){
		println!("The sample size of all combinations based of the size of the given array is less than the size allocated to handle the results of the program.\nKindly, within the program code file edit the sample size constant\n\"const sample_size: usize = {};\"\nto\n\"const sample_size: usize = {};\"\nPreferably a value that is greater than {}.\n", sample_size, sample_space, sample_space); 	
	} else {
		let def_Result: Result = Result::empty();
		let mut h: [Result; number_of_elements] = [def_Result; number_of_elements];
		
		sum_of_three_closest(nums, 1, &mut h);
		
		let mut rn: isize;
		
		unsafe {
			rn = result_number as isize;
		}
		
		for i in 0..rn{
			h[i as usize].print_values();
		}
	}
}
